File: algorithms/02_c/graphs/floyd_warshall.c
/* All-pairs shortest paths (Floyd-Warshall). */
#include <stdio.h>
#define N 4
#define INF 1000000

void floyd_warshall(int dist[N][N]) {
    for (int k = 0; k < N; k++)
        for (int i = 0; i < N; i++)
            for (int j = 0; j < N; j++)
                if (dist[i][k] + dist[k][j] < dist[i][j])
                    dist[i][j] = dist[i][k] + dist[k][j];
}

int main(void) {
    int dist[N][N] = {
        {0, 3, INF, 7},
        {8, 0, 2, INF},
        {5, INF, 0, 1},
        {2, INF, INF, 0},
    };
    floyd_warshall(dist);
    if (dist[0][2] != 5 || dist[0][3] != 6) return 1;
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) printf("%d ", dist[i][j]);
        printf("\n");
    }
    return 0;
}

File: algorithms/11_golang/graphs/floyd_warshall.go
// All-pairs shortest paths (Floyd-Warshall).
package main

import "fmt"

const inf = 1 << 30

func floydWarshall(dist [][]int) [][]int {
	n := len(dist)
	for k := 0; k < n; k++ {
		for i := 0; i < n; i++ {
			for j := 0; j < n; j++ {
				if dist[i][k]+dist[k][j] < dist[i][j] {
					dist[i][j] = dist[i][k] + dist[k][j]
				}
			}
		}
	}
	return dist
}

func main() {
	graph := [][]int{
		{0, 3, inf, 7},
		{8, 0, 2, inf},
		{5, inf, 0, 1},
		{2, inf, inf, 0},
	}
	dist := floydWarshall(graph)
	if dist[0][2] != 5 || dist[0][3] != 6 {
		panic("wrong distance")
	}
	fmt.Println(dist[0])
}
